torch/onnx/_internal/fx/passes/modularization.py

```diff
@@ -0,0 +1,648 @@
+                self.push(_ModuleMeta.from_dynamo_produced_raw_meta(item))
+            else:
+                raise AssertionError(
+                    f"Unknown type of nn_module_stack_meta: {type(raw_meta)}"
+                )
+
+    def __len__(self) -> int:
+        return len(self._module_stack)
+
+    def __getitem__(self, index: int) -> _ModuleMeta:
+        return self._module_stack[index]
+
+    def __iter__(self) -> Iterator[_ModuleMeta]:
+        return iter(self._module_stack)
+
+    def is_empty_or_root(self) -> bool:
+        return len(self._module_stack) == 0
+
+    def top(self) -> _ModuleMeta:
+        """Returns the top module meta in the stack. I.e., the meta for leaf module."""
+        if self.is_empty_or_root():
+            return _ModuleMeta.root()
+        return self._module_stack[-1]
+
+    @_beartype.beartype
+    def is_superset_of(
+        self,
+        module_stack: _ModuleStackMeta,
+    ) -> bool:
+        """Determines if self is a superset of the provided module stack.
+
+        I.e., If self includes all elements from the provided module stack, plus additional
+        elements op top. If self is empty or root, this method always return False.
+        """
+        if self.is_empty_or_root():
+            return False
+
+        if module_stack.is_empty_or_root() is None:
+            return True
+
+        if len(self) <= len(module_stack):
+            return False
+
+        for i, parent_key in enumerate(module_stack):
+            if self[i] != parent_key:
+                return False
+
+        return True
+
+    def push(self, module_meta: _ModuleMeta) -> None:
+        """Pushes a module meta to the stack."""
+        self._module_stack.append(module_meta)
+
+    @_beartype.beartype
+    def is_same(
+        self,
+        module_stack: _ModuleStackMeta,
+    ) -> bool:
+        """Determines if self is the same as the provided module stack."""
+        return self._module_stack == module_stack._module_stack
+
+    @_beartype.beartype
+    def __eq__(self, __value: object) -> bool:
+        if not isinstance(__value, _ModuleStackMeta):
+            return False
+        return self.is_same(__value)
+
+    @property
+    def raw_meta(self) -> Optional[Dict[str, Tuple[str, type]]]:
+        """Returns the raw module stack meta data, i.e. node.meta['nn_module_stack']."""
+        return {
+            module_meta.raw_meta[0]: module_meta.raw_meta[1]
+            for module_meta in self._module_stack
+        }
+
+    def __repr__(self) -> str:
+        return f"ModuleStackMeta({self._module_stack})"
+
+    @property
+    def module_display_name(self) -> str:
+        """Returns the module display name of the top module."""
+        return self.top().module_display_name
+
+    @property
+    def qualified_module_class_name(self) -> str:
+        """Returns the qualified module class name of the top module."""
+        return self.top().qualified_module_class_name
+
+
+@functools.lru_cache()
```

Comment:
Let's not to use cache unless
1. it's a performance bottleneck.
2. we are 100% sure no modification on torch.fx.Node in torch.fx.GraphModule.

Reply:
ok